### PR TITLE
fix accuracy for non-positional rhythm

### PR DESCRIPTION
**Pull request by weixlu**
## Describe the Bug
@liebharc mentioned in #73 that sometimes `eval_pitch_accuracy ` stays at 0.6 to 0.7. I reproduce this issue, and see `eval_pitch_accuracy` stays at 0.67, while `eval_rhythm_accuracy` is correnct: 0.99. This doesn't happen very often, but I do reproduce it several times.

## Find Out What's Wrong
I run the smoke test, and manually check transcription result, where I find something interesting. Although eval_pitch_accuracy is 0.67, I see mostly (>90%) the pitches of notes are correct. So why is eval_pitch_accuracy so low?
Below is a function, which shows we can divide rhythm into two catagory: has position and non posiotion.
```
def has_rhythm_symbol_a_position(rhythm: str) -> bool:
    return rhythm.startswith(("note", "rest", "clef"))
```
- for has-position: yes, the pitch accuracy is high, I have checked via smoke test result
- for non-position: no, the pitch accuracy is low. And **this is indeed where the problem is**.

In this PR, I update `metrics.py` and add this record. From a test training, you can see `eval_pitch_nonposRhy_accuracy` is quite low.
```
eval_pitch_accuracy: 0.679374
eval_pitch_nonposRhy_accuracy: 0.089295
eval_pitch_posRhy_accuracy: 0.997130
eval_rhythm_accuracy: 0.996563
....(other metrics)
```
So why is `eval_pitch_nonposRhy_accuracy` so low?
Let's go to `vocabulary.py`. 
- In `def build_rhythm()`, token id 0 is padding. 
- In `def build_pitch()`, token id 0 is nonote. Nonote, e.g. , there isn't a pitch or note, e.g. , this is a non-position rhythm
So the suspect is, for label `nonote`, the training fails. So `eval_pitch_nonposRhy_accuracy` is low.

We can verify this suspect by looking at `class ScoreDecoder`. This line `self.ignore_index = config.pad_token` is wrong, because `pad_token` is token id 0, and `nonote` is also token id 0. If we let `pad_token` to be ignored, then `nonote` is also ignored. As a result, we see the pitch accuracy for non-position rhythm is low.

## How to Fix
That's easy. Don't let `ignore_index = config.pad_token`, but use `mask` to filter the invalid ones. 
After the fix, I have run the training(10 epochs) for 5 times, and don't see eval_pitch_accuracy to be 0.67 anymore.

## What's Next
- I will run training more times to ensure this won't happen again.
- All the trainings are run on #75 , not on branch main. But I assume this won't make a difference.
- To fix this bug, only code change in `decoder.py` is needed. Those in `metrics.py` are only used to diagnostic problems, and it's OK if we remove them.

### CODE REVIEW

**Comment by liebharc** [comment]
@weixlu thanks once more! That's potentially a really good find. Would you have a fully trained model with this fix, and could you share the pth and trainer state?

File: homr/transformer/vocabulary.py [hunk 1]
```diff
@@ -5,6 +5,8 @@
 from fractions import Fraction
 from typing import Iterable
 
+import torch
```
Comment (liebharc) [inline homr/transformer/vocabulary.py]: We can't use torch from the homr/ folder as pytorch is an optional dependency. So this code should go to training_vocabulary.py.

**Comment by weixlu** [comment]
> @weixlu thanks once more! That's potentially a really good find. Would you have a fully trained model with this fix, and could you share the pth and trainer state?

[https://get.hidrive.com/8bFSW9cQ](https://get.hidrive.com/8bFSW9cQ) This is a model trained for 10 epochs. If it’s not fully trained yet, please let me know how many more epochs are needed, and I’ll resume training.

File: homr/transformer/vocabulary.py (hunk 1 above)
Comment (weixlu) [reply homr/transformer/vocabulary.py]: Ok, I have fixed.

**Comment by liebharc** [comment]
Looks good. Yes, it would be good to see how it performs when training is complete (35 epochs, or early stopping). Thanks!

For this complete run it would be good to not rebase your branches. For homr I always tried to be very traceable with the models. That is why the git commit hash is part of the model name. If we ever find that we went down a wrong road with the model and that a previous model worked much better than it's very easy to see from the model name the exact state of the repo which it was trained on. Also in the past year or so I really tried hard to even commit failed attempts to improve the model and revert them afterward. Because these failures are important learnings and this way it's recorded in the git history what was attempted in the past.

So for your case it would be good to keep your change with https://github.com/liebharc/homr/pull/75 as that's what you are training on.

Thanks again for spotting this issue!

**Comment by weixlu** [comment]
@liebharc OK, I got it. In that case, I'd suggest waiting until #75 is merged before merging this PR.

Regarding #75, is there anything I can do? For example, I'd be happy to run a full training on that branch (with the dataset you send me) and share the model.

**Comment by liebharc** [comment]
#75 mainly needs a training run which gives the same performance as we have on main right now. That's because we have to upload a new model due to the change of vocabulary.

If your change works then we could merge #75 and this one together and upload the training run which had both changes.

**Comment by weixlu** [comment]
> #75 mainly needs a training run which gives the same performance as we have on main right now. That's because we have to upload a new model due to the change of vocabulary.
> 
> If your change works then we could merge #75 and this one together and upload the training run which had both changes.

OK, let me run training on #75 and then share the model

**Comment by weixlu** [comment]
@liebharc Hi, I have finished complete training for both this PR and #75. Smoke test shows SER at 6%-7%. Let me send you `.pth` and `trainer_state.json`.
Somtimes HiDrive is unavailable here in China, so I'll send you a e-mail with the download link from my cloud storage.

**Comment by liebharc** [comment]
@weixlu really thanks a lot for doing this! I've uploaded the model and merged the changes to main. I'm really happy that someone else also looks into the transformer and its training in detail!

**Comment by weixlu** [comment]
> @weixlu really thanks a lot for doing this! I've uploaded the model and merged the changes to main. I'm really happy that someone else also looks into the transformer and its training in detail!

My pleasure :) . I'm really glad the changes were helpful. 

btw, I notice #79 hasn't trained yet. What if I run a full training and share the model again?

**Comment by liebharc** [comment]
Ah I just closed #79. I don't think it's very promising. If you have spare resources then you can give it a go, but it's really not important.

**Comment by weixlu** [comment]
> Ah I just closed #79. I don't think it's very promising. If you have spare resources then you can give it a go, but it's really not important.

OK, then I'll first look into issue #81
